### PR TITLE
match the boundary type with kernel driver

### PR DESCRIPTION
**Pull request by dvdli**
1. Remove calculation of boundary and retrieve boundary from ALSA
kernel. Also fix type when getting available and advance appl_ptr.
2. Fix avail in pcm_mmap_transfer getting overflow.
3. Fix typo.

### CODE REVIEW

File: src/pcm.c [hunk 1]
```diff
@@ -1235,45 +1229,49 @@ int pcm_stop(struct pcm *pcm)
...
 
-static inline int pcm_mmap_capture_avail(struct pcm *pcm)
+static inline long pcm_mmap_capture_avail(struct pcm *pcm)
 {
-    int avail = pcm->mmap_status->hw_ptr - pcm->mmap_control->appl_ptr;
-    if (avail < 0)
+    long avail = pcm->mmap_status->hw_ptr - pcm->mmap_control->appl_ptr;
+    if (avail < 0) {
         avail += pcm->boundary;
+    }
```
Comment (altocoder) [inline src/pcm.c:1251]: from https://android-review.googlesource.com/c/platform/external/tinyalsa/+/949679

    } else if ((unsigned long)avail >= pcm->boundary) {
        avail -= pcm->boundary;
    }

**Comment by philburk** [comment]
This looks good as long as it passes the 13 hour playback tests.
If you add Andy's suggestion then I will approve.

**Comment by dvdli** [comment]
Hi @altocoder and @philburk, 

Thanks a lot for reviewing this PR. I'm thinking of what cases we will encounter when getting avail. In the normal cases, avail is less than buffer size. We don't need to adjust it after calculating, but need to after either hw_ptr or appl_ptr gets wrapped.

On the playback side, we know that appl_ptr is always greater than or equal to hw_ptr, so when appl_ptr is wrapped and hw_ptr is not, we need to take care of the condition `avail >= boundary`.

For example, 
```
boundary = 1024, buffer size = 128, hw_ptr = 1000, appl_ptr = 0
avail = hw_ptr + buffer size - appl_ptr = 1000 + 128 - 0 = 1128 >= boundary.
avail -= boundary => avail = 104
```

I'm curious when we will encounter `avail < 0` on the playback side. Does that mean hw_ptr goes faster than appl_ptr (XRUN)? What does avail mean in XRUN state?

On the capture side, hw_ptr is always greater than or equal to appl_ptr, so when hw_ptr is wrapped, but appl_ptr is not, we need to take care of the condition `avail < 0`.

For example, 
```
boundary = 1024, buffer size = 128, hw_ptr = 0, appl_ptr = 1000
avail = hw_ptr - appl_ptr = 0 - 1000 = -1000 < 0
avail += boundary => avail = 24
```

Note that in ALSA kernel, [snd_pcm_capture_avail](https://elixir.bootlin.com/linux/latest/source/include/sound/pcm.h#L772) tests whether avail < 0 on the capture side. This can also be found in alsa-lib [__snd_pcm_capture_avail](https://github.com/alsa-project/alsa-lib/blob/master/src/pcm/pcm_local.h#L519).

**Comment by philburk** [comment]
@dvdli - Thanks for the explanation. I see that the capture side is not symmetric with playback and the check against boundary is not needed.

> I'm curious when we will encounter avail < 0 on the playback side.

I think that can happen when the hw_ptr has wrapped but the appl_ptr has not. It is not necessarily an XRUN.

File: src/pcm.c (hunk 1 above)
Comment (philburk) [reply src/pcm.c:1251]: I agree with David that this is probably not needed but shouldn't hurt.

**Review by philburk — APPROVED** [review]

**Pull request merged**